### DISCUSSION TITLE
navigation drawer does not work on mobile

### DISCUSSION THREAD

**Discussion by xtream-club** [flet-dev/flet#2070] [General]
Navigation drawer does not work on mobile.  When deploying my application with serious python using navigation drawer it tells me that navigation drawer is not a flutter control and shows an error.  I don't know if it hasn't been tested on android yet or what the problem is.  But it seems that it does not work on mobile devices at the moment.

**Comment by ndonkoHenri** [comment]
Yeah, you're right! Here is what i get:
<img src="https://github.com/flet-dev/flet/assets/98978078/7073e6be-b496-4b07-b36f-9a354b5458e4" height=500>

```python
import flet as ft


def main(page: ft.Page):
    page.drawer = ft.NavigationDrawer(
        controls=[
            ft.Container(height=12),
            ft.NavigationDrawerDestination(
                label="Item 1",
                icon=ft.icons.DOOR_BACK_DOOR_OUTLINED,
                selected_icon_content=ft.Icon(ft.icons.DOOR_BACK_DOOR),
            ),
            ft.Divider(thickness=2),
            ft.NavigationDrawerDestination(
                icon_content=ft.Icon(ft.icons.MAIL_OUTLINED),
                label="Item 2",
                selected_icon=ft.icons.MAIL,
            ),
            ft.NavigationDrawerDestination(
                icon_content=ft.Icon(ft.icons.PHONE_OUTLINED),
                label="Item 3",
                selected_icon=ft.icons.PHONE,
            ),
        ],
    )

    def show_drawer(e):
        page.drawer.open = True
        page.drawer.update()

    page.add(ft.ElevatedButton("Show drawer", on_click=show_drawer))


ft.app(main)
```

**Comment by FeodorFitsner** [comment]
https://github.com/flet-dev/flet/issues/2084

**Comment by ErikEgidoBlanes2223** [comment]
My problem is that if I do it with views, and in a view I add a navigationdrawer I get an error: AssertionError: Control must be added to the page first.

**Reply by ndonkoHenri** [reply]
Please share some code illustrating what you mean.

**Reply by ErikEgidoBlanes2223** [reply]
I have this file main.py 

> 

>      import flet as ft
>      import os
>      from views.page_inicial import Benvinguda
>      from views.page2 import Main2
>      from views.page_info_app import InfoApp
> 
>      ruta_base = os.path.dirname(__file__)
>      carpeta_assets = os.path.join(ruta_base, "/assets")
>      carpeta_fonts = os.path.join(carpeta_assets, "fonts")
>      tipografia_lato = os.path.join(carpeta_fonts,"Lato.ttf")
> 
>      def main(page: ft.Page):
> 
>     page_portada = Benvinguda(page)
>     page2 = Main2(page)
>     pageInfoApp = InfoApp(page)
> 
> 
>     def __init__(self, page: ft.Page):
>         page.title = "App Ajuntament"
>         page.bgcolor="#b1aaa5" #color gris de la pàgina
>         page.fonts = {"Lesson": tipografia_lato} 
>         page.window_max_height= 400
>           
>     
>     def route_change(e):      
>         page.views.clear()
>         page.views.append(
>             ft.View(
>                 "/",
>                 [
>                    page_portada
>                 ],bgcolor="#b1aaa5", scroll=ft.ScrollMode.AUTO
>             )
>         )
>         if page.route == "/home":
>             page.views.append(
>                 ft.View(
>                     "/home",
>                     [
>                         page2
>                     ],bgcolor="#b1aaa5", scroll=ft.ScrollMode.AUTO
>                 )
>             )
>         elif page.route == "/notes":
>             page.views.append(
>                 ft.View(
>                     "/notes",
>                     [
>                         "pageTask"
>                     ],bgcolor="#b1aaa5", scroll=ft.ScrollMode.AUTO
>                 )
>             )
>         elif page.route == "/infoapp":
>             page.views.append(
>                 ft.View(
>                     "/infoapp",
>                     [
>                         pageInfoApp
>                     ],bgcolor="#b1aaa5", scroll=ft.ScrollMode.AUTO
>                 )
>             )
>         page.update()
> 
>     def view_pop(e):
>         print("View pop:", e.view)
>         page.views.pop()
>         top_view = page.views[-1]
>         page.go(top_view.route)
> 
>     page.on_route_change = route_change
>     page.on_view_pop = view_pop
> 
> 
>     page.go(page.route)
> 
> 
> ft.app(target=main)

And i have this page2.py

>           import flet as ft
>          import os
>          import altres.apiDades as apiDades 
>          import altres.db_notes as db_notes
>          from views.page_lateral import MyNavigationDrawer
> 
>          ruta_base = os.path.dirname(__file__)
>          carpeta_assets = os.path.join(ruta_base, "../assets")
> 
>          carpeta_fonts = os.path.join(carpeta_assets, "fonts")
>          carpeta_principal = os.path.join(carpeta_assets, "principal")
>          carpeta_xarxes = os.path.join(carpeta_assets, "xarxes")
> 
>          tipografia_lato = os.path.join(carpeta_fonts,"Lato.ttf")
> 
>          img_correu = os.path.join(carpeta_principal, "correucorporatiu.png")
>          img_crono = os.path.join(carpeta_principal, "fitxar.png")
>          img_empleados = os.path.join(carpeta_principal, "portal_empleado.png")
>          img_ajuntament = os.path.join(carpeta_principal, "web_ajuntament.png")
>          img_seuElectronica = os.path.join(carpeta_principal, "seuElectronica.png")
>          img_seuTributaria = os.path.join(carpeta_principal, "seuTributaria.png")
>          img_gecor = os.path.join(carpeta_principal, "gecor.png")
>          img_extensionesTlf= os.path.join(carpeta_principal,"extensionsTlf.png")
>          img_ajunta= os.path.join(carpeta_principal,"ajunta.png")
>          img_error= os.path.join(carpeta_xarxes,"error.png")
> 
>         try:
>     infoApi = apiDades.omplirDadesActuals()
> 
>     class Main2(ft.Container):
>         def __init__(self, page: ft.Page):
>             super().__init__()
>             page.title = "App Ajuntament"
>             page.bgcolor="#b1aaa5" #color gris de la pàgina
>             page.fonts = {"Lesson": tipografia_lato} 
>             page.window_max_height= 400
> 
> 
>             def obrir_correu(e):
>                 page.launch_url('http://webmail.sueca.es/')
>             def obrir_controlHorari(e):
>                 page.launch_url('https://controlhorari.sueca.es:6443')
>             def obrir_fitxar(e):
>                 page.launch_url('https://controlhorari.sueca.es:6443/fitxar')
>             def obrir_web_sueca(e):
>                 page.launch_url('https://www.sueca.es')
>             def obrir_gecor(e):
>                 page.launch_url("https://portal.gecorcloud.com/login.aspx?a=GECOR")
>             def obrir_seuElectronica(e):
>                 page.launch_url("https://carpetaciutadana.sueca.es/")
>             def obrir_seuTributaria(e):
>                 page.launch_url("https://seutributaria.sueca.es:5443")
>             def obrir_extensionsTlf(e):
>                 page.launch_url("https://drive.google.com/file/d/1dKqzf5lKBq7vgVkpPW04lhoQlDXyJls0/view")
>             def obrir_instagram(e):
>                 page.launch_url("https://www.instagram.com/ajuntamentdesueca/")
>             def obrir_facebook(e):
>                 page.launch_url("https://www.facebook.com/ajuntament.sueca")
>             def obrir_telegram(e):
>                 page.launch_url("https://t.me/s/AjuntamentdeSueca")
>             def obrir_telf_interes(e):
>                 page.launch_url("https://www.sueca.es/val/pagina/telefons-dinteres")
>             def obrir_calendari(e):
>                 page.launch_url("https://carpetaciutadana.sueca.es/opensiac/action/calendariolaboral?method=enter")
>             
>             mynav= MyNavigationDrawer(page=page,infoApi=infoApi,open_calendar=obrir_calendari,open_tlf=obrir_telf_interes,open_insta=obrir_instagram,open_facebook=obrir_facebook,open_telegram=obrir_telegram)
> 
>                   
>             page.drawer = mynav
> 
>             # Funció per a obrir i tancar la barra lateral
>             def show_drawer(e):
>                 page.drawer.open = True
>                 print("holaaaaa obric el panel")
>                 page.drawer.update()   
>                       
>             self.content= ft.Container(content=(
>                             ft.Column(
>                                     controls=[      
>                                         ft.Divider(height=30, opacity=0.0),
>                                         
>                                         ft.IconButton(icon=ft.icons.MENU, on_click=show_drawer, icon_color="white"),
>                                         ft.Divider(height=30, opacity=0.0),
>                                         
>                                         ft.Row(
>                                             controls=[
>                                                 ft.Container(
>                                                     content=(
>                                                         ft.Column(
>                                                             controls=[
>                                                                 ft.Container(content=ft.Image(src=img_correu),width=70,height=70),
>                                                                 ft.Text(value="CORREU CORPORATIU", size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                                             ], 
>                                                             alignment=ft.MainAxisAlignment.CENTER, 
>                                                             horizontal_alignment="center"
>                                                         )
>                                                     ), 
>                                                     width=140,
>                                                     height=140,
>                                                     bgcolor=ft.colors.WHITE, 
>                                                     border_radius=30.0, 
>                                                     on_click=obrir_correu
>                                                 ),
>                                                 
>                                                 ft.Container(
>                                                     content=(
>                                                         ft.Column(
>                                                             controls=[
>                                                                 ft.Container(content=ft.Image(src=img_ajuntament),width=70,height=70),
>                                                                 ft.Text(value="WEB AJUNTAMENT", size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                                             ], 
>                                                             alignment=ft.MainAxisAlignment.CENTER, 
>                                                             horizontal_alignment="center"
>                                                         )
>                                                     ), 
>                                                     width=140,
>                                                     height=140,
>                                                     bgcolor=ft.colors.WHITE, 
>                                                     border_radius=30.0, 
>                                                     on_click=obrir_web_sueca
>                                                 ),                                                                           
>                                             ],
>                                             alignment=ft.MainAxisAlignment.SPACE_EVENLY,
>                                         ),
>                                         
>                                         ft.Row(
>                                             controls=[
>                                                 
>                                                 ft.Container(
>                                                     content=(
>                                                         ft.Column(
>                                                             controls=[
>                                                                 ft.Container(content=ft.Image(src=img_crono),width=70,height=70),
>                                                                 ft.Text(value="FITXATGE MÒBIL", size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                                             ], 
>                                                             alignment=ft.MainAxisAlignment.CENTER, 
>                                                             horizontal_alignment="center"
>                                                         )
>                                                     ), 
>                                                     width=140,
>                                                     height=140,
>                                                     bgcolor=ft.colors.WHITE, 
>                                                     border_radius=30.0, 
>                                                     on_click=obrir_fitxar
>                                                 ),
>                                                 
>                                                 ft.Container(
>                                                     content=(
>                                                         ft.Column(
>                                                             controls=[
>                                                                 ft.Container(content=ft.Image(src=img_empleados),width=70,height=70),
>                                                                 ft.Text(value="PORTAL DEL EMPLEAT", size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                                             ], 
>                                                             alignment=ft.MainAxisAlignment.CENTER, 
>                                                             horizontal_alignment="center"
>                                                         )
>                                                     ), 
>                                                     width=140,
>                                                     height=140,
>                                                     bgcolor=ft.colors.WHITE, 
>                                                     border_radius=30.0, 
>                                                     on_click=obrir_controlHorari
>                                                 ),                                                                                   
>                                             ],
>                                             alignment=ft.MainAxisAlignment.SPACE_EVENLY,
>                                         ),
>                                         
>                                         ft.Row(
>                                             controls=[
>                                                 
>                                                 # QUINT BOTÓ 
>                                                 ft.Container(
>                                                     content=(
>                                                         ft.Column(
>                                                             controls=[
>                                                                 ft.Container(content=ft.Image(src=img_seuElectronica),width=70,height=70),
>                                                                 ft.Text(value="SEU ELECTRÒNICA", size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                                             ], 
>                                                             alignment=ft.MainAxisAlignment.CENTER,
>                                                             horizontal_alignment="center"
>                                                         )
>                                                     ), 
>                                                     width=140,
>                                                     height=140,
>                                                     bgcolor=ft.colors.WHITE, 
>                                                     border_radius=30.0, 
>                                                     on_click=obrir_seuElectronica
>                                                 ),
>                                                 
>                                                 ft.Container(
>                                                     content=(
>                                                         ft.Column(
>                                                             controls=[
>                                                                 ft.Container(content=ft.Image(src=img_seuTributaria),width=70,height=70),
>                                                                 ft.Text(value="SEU TRIBUTARIA", size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                                             ], 
>                                                             alignment=ft.MainAxisAlignment.CENTER, 
>                                                             horizontal_alignment="center"
>                                                         )
>                                                     ), 
>                                                     width=140,
>                                                     height=140,
>                                                     bgcolor=ft.colors.WHITE, 
>                                                     border_radius=30.0, 
>                                                     on_click=obrir_seuTributaria
>                                                 ),                                        
>                                             ],
>                                             alignment=ft.MainAxisAlignment.SPACE_EVENLY,
>                                         ),
>                                         
>                                         ft.Row(
>                                             controls=[
>                                                 
>                                                     ft.Container(
>                                                         content=(
>                                                             ft.Column(
>                                                                 controls=[
>                                                                     ft.Container(content=ft.Image(src=img_extensionesTlf),width=70,height=70),
>                                                                     ft.Text(value="EXTENSIONS TELEFÒN", size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                                                 ], 
>                                                                 alignment=ft.MainAxisAlignment.CENTER, 
>                                                                 horizontal_alignment="center"
>                                                             )
>                                                         ), 
>                                                         width=140,
>                                                         height=140,
>                                                         bgcolor=ft.colors.WHITE, 
>                                                         border_radius=30.0, 
>                                                         on_click=obrir_extensionsTlf
>                                                     ),
>                                                     ft.Container(
>                                                         content=(
>                                                             ft.Column(
>                                                                 controls=[
>                                                                     ft.Container(content=ft.Image(src=img_gecor),width=70,height=70),
>                                                                     ft.Text(value="GECOR ADMINISTRATIUS", size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                                                 ], 
>                                                                 alignment=ft.MainAxisAlignment.CENTER, 
>                                                                 horizontal_alignment="center"
>                                                             )
>                                                         ), 
>                                                         width=140,
>                                                         height=140,
>                                                         bgcolor=ft.colors.WHITE, 
>                                                         border_radius=30.0, 
>                                                         on_click=obrir_gecor
>                                                     ),                                        
>                                             ],
>                                             alignment=ft.MainAxisAlignment.SPACE_EVENLY,
>                                         ),
>                                     ])
>                             ),bgcolor="#b1aaa5")
>             
> 
>          except Exception as e:
>     
>      class Main2(ft.Container):
>         def __init__(self, page: ft.Page):
>             super().__init__()
>             page.title = "App Ajuntament"
>             page.bgcolor="#b1aaa5"
> 
>             self.content= ft.Column(controls=[
>                 ft.Divider(height=200, opacity=0.0),
>                 ft.Row(controls=[
>                     ft.Container(
>                                 content=(
>                                     ft.Column(
>                                         controls=[
>                                             ft.Container(content=ft.Image(src=img_error),width=70,height=70),
>                                             ft.Text(value="REVISA LA TEUA CONEXIÓ A INTERNET\n TANCA L'APLICACIÓ I TORNA-HO A INTENTAR.", 
>                                                     size=10,color=ft.colors.BLACK, weight=ft.FontWeight.BOLD, text_align=ft.TextAlign.CENTER),
>                                         ], 
>                                         alignment=ft.MainAxisAlignment.SPACE_EVENLY, 
>                                         horizontal_alignment="center"
>                                     )
>                                 ), 
>                                 width=300,
>                                 height=200,
>                                 bgcolor=ft.colors.WHITE, 
>                                 border_radius=30.0, 
>                             )
>                     ],alignment="center",vertical_alignment="center")
>             ])
> 
>              
>                 
>          if __name__ == "__main__":
>              def main(page: ft.Page):
>                  page.add(Main2(page))
>              ft.app(target=main)

**Reply by ErikEgidoBlanes2223** [reply]
AssertionError: Control must be added to the page first.

**Reply by ErikEgidoBlanes2223** [reply]
page_portada is a single page with a button that goes to page2

**Reply by ndonkoHenri** [reply]
Can it be possible for you to come up with a simple example reproducing the issue? 
What you shared seems complex for quick review/testing.

**Reply by baseplate-admin** [reply]
Perhaps a github repository would be better in this case?

@ErikEgidoBlanes2223. could you please put this in a github repository?

**Reply by ErikEgidoBlanes2223** [reply]
[Uploading error-flet.zip…]()

**Reply by baseplate-admin** [reply]
You would have to wqit a bit longer. After uploading

**Reply by ErikEgidoBlanes2223** [reply]
As? I do not understand friend

**Reply by baseplate-admin** [reply]
Uh the zip file is not downloadable.. 

You have to wait a bit longer till it actually finishes being uploaded into github. The text should change from 'Uploading' to something else

**Reply by ErikEgidoBlanes2223** [reply]
[https://drive.google.com/drive/folders/1l5y7gZeJltsZZY7znD3wE6uZMe22jXoB?usp=sharing](url)

I add the drive link here so you can see the video, since Github gives me an error when uploading it.

**Reply by ErikEgidoBlanes2223** [reply]
[ndonkoHenri](https://github.com/ndonkoHenri)

[https://drive.google.com/drive/folders/1l5y7gZeJltsZZY7znD3wE6uZMe22jXoB?usp=sharing](https://github.com/flet-dev/flet/discussions/url)

I add the drive link here so you can see the video, since Github gives me an error when uploading it.

**Reply by baseplate-admin** [reply]
[ajunt-20240510T113327Z-001.zip](https://github.com/flet-dev/flet/files/15274484/ajunt-20240510T113327Z-001.zip)

https://github.com/flet-dev/flet/assets/61817579/084a1a11-7e55-47ae-9654-46a251754553

**Reply by ErikEgidoBlanes2223** [reply]
Friend, did you find any solution?

**Reply by baseplate-admin** [reply]
Unfortunately nope,  i didn’t have time to really dig into this issue. 

I will try to, but there might be others who are faster/effective than me

**Reply by ErikEgidoBlanes2223** [reply]
Hello friend, I already found exactly where I have the problem, I am creating a view, I pass a container value to it, and within that container I have all the characteristics of my page, the error is that if I put my navigationdrawer in the container that then it will be inside the view gives me the error experienced.

However, if I add a drawer option to my view I can use it.

Now my problem is that I don't know what to put in the drawer option of ft.View, and I don't know how to add the functionality for when it is clicked and so on.

**Reply by ErikEgidoBlanes2223** [reply]
```if page.route == "/home":
            page.views.append(
                ft.View(
                    "/home",
                    [
                        page2
                    ],bgcolor="#b1aaa5", scroll=ft.ScrollMode.AUTO, drawer= MyNavigationDrawer
                )
            )
```

page2 is a object ft.Container, but ft.View I have option drawer.

**Reply by tinyabdu** [reply]
Bro I have same issue if I use view

**Reply by tinyabdu** [reply]
> ```if page.route == "/home":
>             page.views.append(
>                 ft.View(
>                     "/home",
>                     [
>                         page2
>                     ],bgcolor="#b1aaa5", scroll=ft.ScrollMode.AUTO, drawer= MyNavigationDrawer
>                 )
>             )
> ```
> 
> page2 is a object ft.Container, but ft.View I have option drawer.  

Bro can you explain more
